transaction/std/src/memo/authenticated_sender_with_payment_request_id.rs

```diff
@@ -0,0 +1,160 @@
+
+        let shared_secret = cred
+            .subaddress_spend_private_key
+            .key_exchange(receiving_subaddress_view_public_key);
+
+        let hmac_value = compute_category1_hmac(
+            shared_secret.as_ref(),
+            tx_out_public_key,
+            Self::MEMO_TYPE_BYTES,
+            &memo_data,
+        );
+        memo_data[28..].copy_from_slice(&hmac_value);
+
+        Self { memo_data }
+    }
+
+    /// Get the sender address hash from the memo
+    pub fn sender_address_hash(&self) -> AddressHash {
+        let bytes: [u8; 16] = self.memo_data[0..16].try_into().unwrap();
+        AddressHash::from(bytes)
+    }
+
+    /// Get the payment request id from the memo
+    pub fn payment_request_id(&self) -> u64 {
+        u64::from_be_bytes(self.memo_data[16..24].try_into().unwrap())
+    }
+
+    /// Validate an AuthenticatedSenderMemo
+    ///
+    /// First, the client should look up the sender's Public Address from their
+    /// hash. If it isn't a known contact we won't be able to authenticate
+    /// them.
+    ///
+    /// Then they need to get the view private key corresponding to the
+    /// subaddress that this TxOut was sent to. This is usually our default
+    /// subaddress view private key.
+    ///
+    /// Finally we can validate the memo against these data. The
+    /// tx_out_public_key is also under the mac, which prevents replay
+    /// attacks.
+    ///
+    /// Arguments:
+    /// * sender_address: The public address of the sender. This can be looked
+    ///   up by the AddressHash provided.
+    /// * receiving_subaddress_view_private_key: This is usually our
+    ///   default_subaddress_view_private_key, but should correspond to whatever
+    ///   subaddress recieved this TxOut.
+    /// * tx_out_public_key: The public key of the TxOut to which this memo is
+    ///   attached.
+    ///
+    /// Returns:
+    /// * subtle::Choice(1u8) if validation passed, subtle::Choice(0u8) if hmac
+    ///   comparison failed.
+    ///
+    /// This function is constant-time.
+    pub fn validate(
+        &self,
+        sender_address: &PublicAddress,
+        receiving_subaddress_view_private_key: &RistrettoPrivate,
+        tx_out_public_key: &CompressedRistrettoPublic,
+    ) -> Choice {
+        let mut result = Choice::from(1u8);
+        let sender_address_hash = AddressHash::from(sender_address);
+        result &= sender_address_hash.ct_eq(&self.sender_address_hash());
+
+        let shared_secret =
+            receiving_subaddress_view_private_key.key_exchange(sender_address.spend_public_key());
+
+        let expected_hmac = compute_category1_hmac(
+            shared_secret.as_ref(),
+            tx_out_public_key,
+            Self::MEMO_TYPE_BYTES,
+            &self.memo_data,
+        );
+        let found_hmac: [u8; 16] = self.memo_data[28..].try_into().unwrap();
+        result &= expected_hmac.ct_eq(&found_hmac);
+        result
```

Comment:
This seems like copy-paste from `AuthenticatedSender`, can this be de-duped (e.g. type-parameterized helper function, with `T::MEMO_TYPE_BYTES`)?

Reply:
yeah seems reasonable, will do